src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,224 @@
+      store_provider_(object_interface.CreateStoreProvider(StoreProviderType::LOCAL_PLASMA)),
+      counter_(0) {
+       
+  RAY_CHECK_OK(SubscribeActorTable());
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(const TaskSpec &task) {
+
+  if (HasByReferenceArgs(task.GetTaskSpecification())) {
+    return Status::Invalid("direct actor call only supports by value arguments");
+  }
+
+  RAY_CHECK(task.GetTaskSpecification().IsActorTask());
+  const auto &actor_id = task.GetTaskSpecification().ActorId();
+
+  const auto task_id = task.GetTaskSpecification().TaskId();
+  const auto num_returns = task.GetTaskSpecification().NumReturns();
+
+  auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
+  request->set_task_id(task.GetTaskSpecification().TaskId().Binary());
+  request->set_task_spec(task.GetTaskSpecification().Serialize());
+
+  std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+  auto entry = rpc_clients_.find(actor_id);
+  if (entry == rpc_clients_.end()) {
+    // TODO: what if actor is not created yet?
+    // if (pending_requests_[actor_id].empty()) {
+    //   gcs_client_.Actors().RequestNotifications(JobID::Nil(), actor_id,
+    //      gcs_client_.GetLocalClientID());
+    // }
+
+    auto pending_request =
+        std::unique_ptr<PendingTaskRequest>(new PendingTaskRequest);
+    pending_request->task_id = task_id;
+    pending_request->num_returns = num_returns;
+    pending_request->request = std::move(request);
+    // append the task to the pending list.
+
+    pending_requests_[actor_id].emplace_back(std::move(pending_request));
+
+    return Status::OK();
+  }
+
+  auto iter = actor_state_.find(actor_id);
+  RAY_CHECK(iter != actor_state_.end());
+  if (iter->second != ActorTableData::ALIVE) {
+    TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    return Status::IOError("actor is dead or being reconstructed");
+  }
+
+  auto &client = entry->second;
+  RAY_LOG(DEBUG) << "push task " << "   " << task_id << "    " << client.get();  
+  auto status = PushTask(*client, *request, task_id, num_returns);
+
+  return status;
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubscribeActorTable() {
+  // Register a callback to handle actor notifications.
+  auto actor_notification_callback = [this](const ActorID &actor_id,
+                                            const std::vector<ActorTableData> &data) {
+    if (!data.empty()) {
+      const auto &actor_data = data.back();
+      
+      if (actor_data.state() == ActorTableData::ALIVE) {
+        RAY_LOG(INFO) << "received notification on actor alive, actor_id: " << actor_id
+                      << ", ip address: " << actor_data.ip_address()
+                      << ", port: " << actor_data.port();
+
+        std::unique_ptr<rpc::DirectActorClient> grpc_client(
+            new rpc::DirectActorClient(actor_data.ip_address(),
+            actor_data.port(), client_call_manager_));
```

Comment:
I think we shouldn't establish connections for all actors. There will be too many. We can do that on demand.

Reply:
right... originally I was using `RequestNotification` to choose interested actors, but that isn't available at the first delivery of `GcsClient`, a possible way is to track by other data structures.

Reply:
updated.

Reply:
I think there are 2 options:
1) Subscribe all actors, don't establish connections immediately but only save actor info in a local map. Establish connections on demand when submitting actor tasks.
2) Don't subscribe anything initially. When receiving a new actor task,  look up actor info in gcs and establish a connection. And then also subscribe that particular actor.

Option 1 should have shorter latency, but option 2 can subscribe less data. I'm not too sure which one would be better. They both sound fine to me.

BTW, we can add a method in `ActorStateAccessor` that subscribes a given actor. That change should be simple, because we only need to do `RequestNotification` in that method.

Reply:
Thanks, I'm taking option 1 for now, as the interface to subscribe a specific actor is not yet available.